lib/chef/application/exit_code.rb

```diff
@@ -0,0 +1,218 @@
+#
+# Author:: Steven Murawski (<smurawski@chef.io>)
+# Copyright:: Copyright 2016, Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+class Chef
+  class Application
+
+    # These are the exit codes defined in Chef RFC 062
+    # https://github.com/chef/chef-rfc/blob/master/rfc062-exit-status.md
+    class ExitCode
+
+      # -1 is defined as DEPRECATED_FAILURE in RFC 062, so it is
+      # not enumerated in an active constant.
+      #
+      VALID_RFC_062_EXIT_CODES = {
+        SUCCESS: 0,
+        GENERIC_FAILURE: 1,
+        SIGINT_RECEIVED: 2,
+        SIGTERM_RECEIVED: 3,
+        REBOOT_SCHEDULED: 35,
+        REBOOT_NEEDED: 37,
+        REBOOT_FAILED: 41,
+        AUDIT_MODE_FAILURE: 42,
+      }
+
+      DEPRECATED_RFC_062_EXIT_CODES = {
+        DEPRECATED_FAILURE: -1,
+      }
+
+      class << self
+
+        def validate_exit_code(exit_code = nil)
+          exit_code = resolve_exit_code(exit_code)
+          return exit_code if valid?(exit_code)
+          default_exit_code
+        end
+
+        def valid?(exit_code)
+          return false if exit_code.nil?
+          return true if skip_validation
+          return true if valid_exit_codes.include? exit_code
+
+          notify_on_deprecation(deprecation_warning)
+          allow_deprecated_exit_code
+        end
+
+        def allow_deprecated_exit_code
+          ## After the new exit code behavior default changes
+          ## TODO: change the check to just
```

Comment:
I believe we'll want to announce the new optional behavior, then after a few releases we can default it.  Doing the todo now will start issuing deprecation warnings on current exit behavior.